File: missing-customer-tax-ids-b2b-eu/node/stripe-eu-vat-ids.mjs
/**
 * Report EU business invoices billed with VAT because no tax ID was on file.
 *
 * Read only. Paginated GETs and no writes: give this a RESTRICTED key with read
 * access to Invoices and Customers. The repair is printed, never performed.
 */
const API = 'https://api.stripe.com/v1';

// The 27 member states. Reverse charge is an intra-EU mechanism, so a country
// outside this set is a different question with a different answer.
export const EU = new Set(('AT BE BG HR CY CZ DK EE FI FR DE GR HU IE IT LV LT ' +
  'LU MT NL PL PT RO SK SI ES SE').split(' '));

// Verification results that are not a confirmation. `pending` is normal for a
// few minutes and a problem after a few months.
export const UNCONFIRMED = ['unverified', 'unavailable', 'pending'];

/**
 * Classify one paid invoice. Pure, so the rules can be tested without a network.
 * `invoiceTaxIds` is the invoice's customer_tax_ids array, frozen at finalization.
 */
export function verdict(country, invoiceTaxIds, taxExempt, taxAmount, verification) {
  if (!EU.has(country)) {
    return ['out-of-scope',
      `${country || 'no country on the invoice'} is outside the EU: the reverse ` +
      'charge does not apply here'];
  }
  if (taxExempt === 'reverse') {
    return ['reverse-charge',
      'billed under the reverse charge; the buyer accounts for the VAT'];
  }
  if (taxExempt === 'exempt') {
    return ['exempt', 'recorded as exempt, so no VAT was due and none was charged'];
  }
  if (!invoiceTaxIds || invoiceTaxIds.length === 0) {
    if (taxAmount) {
      return ['charged-vat',
        `no customer_tax_ids on the invoice and ${taxAmount} in tax charged: a ` +
        'business was billed as a consumer'];
    }
    return ['no-id-no-vat',
      'no tax ID and no VAT either; that is a registration question rather than ' +
      'a reverse charge one'];
  }
  if (UNCONFIRMED.includes(verification)) {
    return ['unverified',
      `a tax ID is on the invoice but its verification status is ` +
      `${JSON.stringify(verification)}: not a number to rely on`];
  }
  return ['ok', 'a verified tax ID is recorded on the invoice'];
}

/** Total tax on the invoice in minor units, across every tax line. */
export function taxCharged(inv) {
  return (inv.total_taxes ?? []).reduce((a, t) => a + (t.amount ?? 0), 0);
}

async function get(key, path, params = {}) {
  const url = new URL(API + path);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${key}` } });
  if (res.status === 401) {
    throw new Error('401 from Stripe: the key is wrong, or is for the other mode');
  }
  if (!res.ok) throw new Error(`${res.status} from ${url.pathname}`);
  return res.json();
}

export async function paidInvoices(key, days = 180, limit = 5000) {
  const cutoff = Math.floor(Date.now() / 1000 - days * 86400);
  const out = [];
  const params = { status: 'paid', limit: 100, 'created[gte]': cutoff };
  for (;;) {
    const page = await get(key, '/invoices', params);
    const data = page.data ?? [];
    out.push(...data);
    if (data.length === 0 || !page.has_more || out.length >= limit) break;
    params.starting_after = data[data.length - 1].id;
  }
  return out;
}

async function verificationStatus(key, customerId, cache) {
  if (cache.has(customerId)) return cache.get(customerId);
  let status = null;
  try {
    const { data = [] } = await get(key, `/customers/${customerId}/tax_ids`, { limit: 10 });
    for (const tid of data) {
      const s = tid.verification?.status;
      if (UNCONFIRMED.includes(s)) { status = s; break; }
      status = status ?? s ?? null;
    }
  } catch {
    status = null;
  }
  cache.set(customerId, status);
  return status;
}

async function main() {
  const key = (process.env.STRIPE_API_KEY || "dummy-stripe-api-key");
  if (!key) {
    console.error('set STRIPE_API_KEY (use a restricted, read-only key)');
    process.exitCode = 2;
    return;
  }

  const days = Number(process.argv[2] ?? 180);
  const cache = new Map();
  let euSeen = 0;
  const findings = [];

  for (const inv of await paidInvoices(key, days)) {
    const country = inv.customer_address?.country ?? '';
    if (!EU.has(country)) continue;
    euSeen += 1;
    const cus = (typeof inv.customer === 'object' && inv.customer !== null
      ? inv.customer.id : inv.customer) ?? null;
    const ids = inv.customer_tax_ids ?? [];
    const verification = ids.length && cus
      ? await verificationStatus(key, cus, cache) : null;
    const tax = taxCharged(inv);
    const [state, detail] = verdict(country, ids, inv.customer_tax_exempt, tax, verification);
    if (['charged-vat', 'unverified', 'no-id-no-vat'].includes(state)) {
      findings.push({ state, id: inv.id ?? '<no id>', cus, country, tax, detail });
    }
  }

  if (findings.length === 0) {
    console.log(`${'clear'.padEnd(13)} 0 of ${euSeen} EU invoice(s) billed to a business as a consumer`);
    return;
  }

  const charged = findings.filter((f) => f.state === 'charged-vat');
  const chargedTax = charged.reduce((a, f) => a + f.tax, 0);
  console.warn(`${'no-tax-id'.padEnd(13)} ${findings.length} of ${euSeen} EU invoice(s) flagged, ${charged.length} charged VAT with no tax ID, ${chargedTax} in minor units`);

  findings.sort((a, b) => b.tax - a.tax);
  for (const f of findings.slice(0, 20)) {
    console.warn(`  ${f.state.padEnd(13)} ${f.id}  ${f.cus}  ${f.country}  ${f.tax}  ${f.detail}`);
    if (f.state === 'charged-vat') {
      console.warn(`      repair: POST ${API}/tax_ids  type=eu_vat  value=${f.country}123456789  owner[type]=customer  owner[customer]=${f.cus}`);
      console.warn('      the invoice itself is frozen: correct it with a credit note and a reissue, not an edit');
    }
  }
  if (findings.length > 20) console.warn(`  ... and ${findings.length - 20} more`);
  console.warn('  then switch on tax ID collection in Checkout and allow the ' +
               'tax_id field in the billing portal, or the list refills');
  process.exitCode = 1;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing key, and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}

File: missing-customer-tax-ids-b2b-eu/python/stripe_eu_vat_ids.py
"""Report EU business invoices billed with VAT because no tax ID was on file.

Read only. Paginated GETs and no writes: give this a RESTRICTED key with read
access to Invoices and Customers. The repair is printed, never performed, because
a credit note is a legal record and reissuing an invoice rebills a customer.
"""
import argparse
import logging
import os
import sys
import time

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("stripe_eu_vat_ids")

API = "https://api.stripe.com/v1"

# The 27 member states. Reverse charge is an intra-EU mechanism, so a country
# outside this set is a different question with a different answer.
EU = frozenset("AT BE BG HR CY CZ DK EE FI FR DE GR HU IE IT LV LT LU MT NL PL "
               "PT RO SK SI ES SE".split())

# Verification results that are not a confirmation. `pending` is normal for a few
# minutes and a problem after a few months.
UNCONFIRMED = ("unverified", "unavailable", "pending")


def verdict(country, invoice_tax_ids, tax_exempt, tax_amount, verification):
    """Classify one paid invoice. Pure, so the rules can be tested without a network.

    `invoice_tax_ids` is the invoice's customer_tax_ids array, frozen at
    finalization. `tax_exempt` is customer_tax_exempt. `tax_amount` is the tax
    actually charged in minor units. `verification` is the status of the
    customer's tax ID, or None when there is not one. Returns (state, detail).
    """
    if country not in EU:
        return ("out-of-scope",
                "%s is outside the EU: the reverse charge does not apply here"
                % (country or "no country on the invoice",))
    if tax_exempt == "reverse":
        return ("reverse-charge",
                "billed under the reverse charge; the buyer accounts for the VAT")
    if tax_exempt == "exempt":
        return ("exempt",
                "recorded as exempt, so no VAT was due and none was charged")
    if not invoice_tax_ids:
        if tax_amount:
            return ("charged-vat",
                    "no customer_tax_ids on the invoice and %d in tax charged: a "
                    "business was billed as a consumer" % tax_amount)
        return ("no-id-no-vat",
                "no tax ID and no VAT either; that is a registration question "
                "rather than a reverse charge one")
    if verification in UNCONFIRMED:
        return ("unverified",
                "a tax ID is on the invoice but its verification status is %r: "
                "not a number to rely on" % (verification,))
    return ("ok", "a verified tax ID is recorded on the invoice")


def get(session, path, **params):
    r = session.get(API + path, params=params, timeout=30)
    if r.status_code == 401:
        raise SystemExit("401 from Stripe: the key is wrong, or is for the other mode")
    r.raise_for_status()
    return r.json()


def paid_invoices(session, days, limit):
    """Page paid invoices created within the window."""
    cutoff = int(time.time() - days * 86400)
    out = []
    params = {"status": "paid", "limit": 100, "created[gte]": cutoff}
    while True:
        page = get(session, "/invoices", **params)
        data = page.get("data", [])
        out.extend(data)
        if not data or not page.get("has_more") or len(out) >= limit:
            break
        params["starting_after"] = data[-1]["id"]
    return out


def verification_status(session, customer_id, cache):
    """Weakest verification status across a customer's tax IDs, or None."""
    if customer_id in cache:
        return cache[customer_id]
    status = None
    try:
        ids = get(session, "/customers/%s/tax_ids" % customer_id,
                  limit=10).get("data", [])
    except requests.HTTPError:
        ids = []
    for tid in ids:
        s = (tid.get("verification") or {}).get("status")
        if s in UNCONFIRMED:
            status = s
            break
        status = status or s
    cache[customer_id] = status
    return status


def tax_charged(inv):
    """Total tax on the invoice in minor units, across every tax line."""
    return sum(t.get("amount") or 0 for t in (inv.get("total_taxes") or []))


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--days", type=float, default=180,
                    help="how far back to read paid invoices")
    ap.add_argument("--top", type=int, default=20,
                    help="how many invoices to print")
    ap.add_argument("--max-invoices", type=int, default=5000,
                    help="stop paginating after this many invoices")
    args = ap.parse_args()

    key = os.environ.get("STRIPE_API_KEY")
    if not key:
        log.error("set STRIPE_API_KEY (use a restricted, read-only key)")
        return 2

    s = requests.Session()
    s.headers.update({"Authorization": "Bearer " + key})

    cache = {}
    eu_seen = 0
    findings = []
    for inv in paid_invoices(s, args.days, args.max_invoices):
        country = ((inv.get("customer_address") or {}).get("country") or "")
        if country not in EU:
            continue
        eu_seen += 1
        cus = inv.get("customer")
        if isinstance(cus, dict):
            cus = cus.get("id")
        ids = inv.get("customer_tax_ids") or []
        verification = verification_status(s, cus, cache) if ids and cus else None
        state, detail = verdict(country, ids, inv.get("customer_tax_exempt"),
                                tax_charged(inv), verification)
        if state in ("charged-vat", "unverified", "no-id-no-vat"):
            findings.append((state, inv.get("id", "<no id>"), cus, country,
                             tax_charged(inv), detail))

    if not findings:
        log.info("%-13s 0 of %d EU invoice(s) billed to a business as a consumer",
                 "clear", eu_seen)
        return 0

    charged = [f for f in findings if f[0] == "charged-vat"]
    log.warning("%-13s %d of %d EU invoice(s) flagged, %d charged VAT with no "
                "tax ID, %d in minor units", "no-tax-id", len(findings), eu_seen,
                len(charged), sum(f[4] for f in charged))

    findings.sort(key=lambda f: -f[4])
    for state, inv_id, cus, country, tax, detail in findings[:args.top]:
        log.warning("  %-13s %s  %s  %s  %d  %s",
                    state, inv_id, cus, country, tax, detail)
        if state == "charged-vat":
            log.warning("      repair: POST %s/tax_ids  type=eu_vat  "
                        "value=%s123456789  owner[type]=customer  owner[customer]=%s",
                        API, country, cus)
            log.warning("      the invoice itself is frozen: correct it with a "
                        "credit note and a reissue, not an edit")
    if len(findings) > args.top:
        log.warning("  ... and %d more", len(findings) - args.top)
    log.warning("  then switch on tax ID collection in Checkout and allow the "
                "tax_id field in the billing portal, or the list refills")
    return 1


if __name__ == "__main__":
    sys.exit(main())
